llm-models/llamav2/llamav2-7b/06_fine_tune_qlora_marketplace.py

```diff
@@ -0,0 +1,423 @@
+
+# COMMAND ----------
+
+import mlflow
+
+# Set mlflow registry to databricks-uc
+mlflow.set_registry_uri("databricks-uc")
+
+# COMMAND ----------
+
+catalog_name = "databricks_llama_2_models" # Default catalog name when installing the model from Databricks Marketplace
+version = 1
+
+#model_mlflow_path = f"models:/{catalog_name}.models.llama_2_7b_hf/{version}"
+model_mlflow_path = f"models:/databricks_marketplace.models_lu.llama_2_7b_hf/{version}"
+
+model_local_path = "/local_disk0/llama_2_7b_hf/"
+model_output_local_path = "/local_disk0/llama-2-7b-lora-fine-tune"
+
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Dataset
+# MAGIC
+# MAGIC We will use the [databricks-dolly-15k ](https://huggingface.co/datasets/databricks/databricks-dolly-15k) dataset.
+
+# COMMAND ----------
+
+from datasets import load_dataset
+
+dataset_name = "databricks/databricks-dolly-15k"
+dataset = load_dataset(dataset_name, split="train")
+
+# COMMAND ----------
+
+INTRO_BLURB = "Below is an instruction that describes a task. Write a response that appropriately completes the request."
+INSTRUCTION_KEY = "### Instruction:"
+INPUT_KEY = "Input:"
+RESPONSE_KEY = "### Response:"
+END_KEY = "### End"
+
+PROMPT_NO_INPUT_FORMAT = """{intro}
+
+{instruction_key}
+{instruction}
+
+{response_key}
+{response}
+
+{end_key}""".format(
+  intro=INTRO_BLURB,
+  instruction_key=INSTRUCTION_KEY,
+  instruction="{instruction}",
+  response_key=RESPONSE_KEY,
+  response="{response}",
+  end_key=END_KEY
+)
+
+PROMPT_WITH_INPUT_FORMAT = """{intro}
+
+{instruction_key}
+{instruction}
+
+{input_key}
+{input}
+
+{response_key}
+{response}
+
+{end_key}""".format(
+  intro=INTRO_BLURB,
+  instruction_key=INSTRUCTION_KEY,
+  instruction="{instruction}",
+  input_key=INPUT_KEY,
+  input="{input}",
+  response_key=RESPONSE_KEY,
+  response="{response}",
+  end_key=END_KEY
+)
+
+def apply_prompt_template(examples):
+  instruction = examples["instruction"]
+  response = examples["response"]
+  context = examples.get("context")
+
+  if context:
+    full_prompt = PROMPT_WITH_INPUT_FORMAT.format(instruction=instruction, response=response, input=context)
+  else:
+    full_prompt = PROMPT_NO_INPUT_FORMAT.format(instruction=instruction, response=response)
+  return { "text": full_prompt }
+
+dataset = dataset.map(apply_prompt_template)
+
+# COMMAND ----------
+
+dataset["text"][0]
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Loading the model
+# MAGIC
+# MAGIC In this section we will load the [LLaMAV2](), quantize it in 4bit and attach LoRA adapters on it.
+
+# COMMAND ----------
+
+from mlflow.artifacts import download_artifacts
+
+path = download_artifacts(artifact_uri=model_mlflow_path, dst_path=model_local_path)
+
+# COMMAND ----------
+
+import os
+import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig, AutoTokenizer
+
+tokenizer_path = os.path.join(path, "components", "tokenizer")
+model_path = os.path.join(path, "model")
```

Comment:
This is pretty cool, thanks for demonstrating how to access the model and tokenizer from MLflow!